tests/udp.rs

```diff
@@ -26,6 +27,124 @@ fn udp_peers_not_connected_huge_data_exchange() {
+                        let res: Option<(Vec<u8>, _)> = unwrap!(sock1.read_frm());
+                        if let Some((_data, peer_addr)) = res {
+                            received_from.push(peer_addr);
+                        } else {
+                            break;
+                        }
+                    }
+                    if received_from.len() == 2 {
+                        break 'event_loop;
+                    }
+                }
+                SOCKET2_TOKEN => {
+                    let data_sent =
+                        unwrap!(sock2.write_to(Some((vec![1u8, 2, 3], server_addr, 1))));
+                    assert!(data_sent);
+                    unwrap!(poll.deregister(&sock2));
+                }
+                SOCKET3_TOKEN => {
+                    let data_sent =
+                        unwrap!(sock3.write_to(Some((vec![1u8, 2, 3], server_addr, 1))));
+                    assert!(data_sent);
+                    unwrap!(poll.deregister(&sock3));
+                }
+                _ => panic!("Unexpected event"),
+            }
+        }
+    }
+
+    assert_that!(
+        &received_from,
+        contains(vec![sock2_addr, sock3_addr]).exactly()
+    );
+}
+
+#[test]
+fn connected_socket_receives_packets_only_from_connected_socket() {
+    const SOCKET1_TOKEN: Token = Token(0);
+    const SOCKET2_TOKEN: Token = Token(1);
+    const SOCKET3_TOKEN: Token = Token(2);
+    const TIMEOUT_TOKEN: Token = Token(3);
+
+    let mut sock2 = unwrap!(UdpSock::bind(&unwrap!("127.0.0.1:0".parse())));
+    let sock2_addr = unwrap!(sock2.local_addr());
+
+    let mut sock1 = unwrap!(UdpSock::bind(&unwrap!("127.0.0.1:0".parse())));
+    unwrap!(sock1.connect(&sock2_addr));
+    let server_addr = unwrap!(sock1.local_addr());
+    unwrap!(sock2.connect(&server_addr));
+    let mut sock3 = unwrap!(UdpSock::bind(&unwrap!("127.0.0.1:0".parse())));
+
+    let poll = unwrap!(Poll::new());
+    unwrap!(poll.register(&sock1, SOCKET1_TOKEN, Ready::readable(), PollOpt::edge()));
+    unwrap!(poll.register(&sock2, SOCKET2_TOKEN, Ready::writable(), PollOpt::edge()));
+    unwrap!(poll.register(&sock3, SOCKET3_TOKEN, Ready::writable(), PollOpt::edge()));
+
+    let mut timer = Timer::default();
+    timer.set_timeout(Duration::from_secs(1), ()); // let's terminate the test after 1 second
+    unwrap!(poll.register(&timer, TIMEOUT_TOKEN, Ready::readable(), PollOpt::edge(),));
+
+    let mut received_from = HashSet::new();
+    let mut events = Events::with_capacity(1024);
+    'event_loop: loop {
+        let _ = unwrap!(poll.poll(&mut events, None));
+        for event in events.iter() {
+            match event.token() {
+                SOCKET1_TOKEN => loop {
+                    let res: Option<(Vec<u8>, _)> = unwrap!(sock1.read_frm());
+                    if let Some((_data, peer_addr)) = res {
+                        received_from.insert(peer_addr);
+                    } else {
+                        break;
```

Comment:
why break out of the loop ? If it was a spurious wake and resulted in a WB error we want to continue, not break.

Reply:
On the contrary, if it was `WouldBlock`, then we want to wait when the socket becomes readable again :)

Reply:
meaning we call `read_frm()` until it returns `None`

Reply:
Ya, what i meant was why break and not read the rest of the events in the event-set ? Other sockets could have their readiness triggered too.

Reply:
it only break the inner loop on line 121 :)

Reply:
Ahhhhh there's one more loop - i missed that.